### PR TITLE
Support Apple universal builds

### PR DESCRIPTION
**Pull request by oskirby**
## Description
Compiled targets for Apple devices don't always have a single architecture that they may need to build for. For example MacOS universal binaries will include symbols for both `aarch64` and `x86_64` within a single binary, and iOS projects may need to build libraries for `x86_64` during simulation and `aarch64` for real devices.

For the most part, Qt and CMake has already sorted this out for us, but our custom cmake helpers to bring in Rust and Golang libraries doesn't support building of universaly binaries. This PR is mostly about updating the Golang and Rust helpers to produce universal binaries when appropriate. As a high-level tour of the easy changes in this PR to make this work:
- We added a post-processing script `ios-arch-fixup.cmake` to set some special Xcode attributes for all compiled iOS targets and setup a split architecture build. This basically just tells Xcode that the simulator needs to build for `x86_64` and real iOS devices need to build for `aarch64`
- The iOS network extension now builds the Wireguard module for both `x86_64` and `aarch64`, glues them together into a fat binary with `lipo`. The unusued architecture should get dropped by the Clang linker.
- To enable per-architecture builds, a new golang helper has been added: `build_go_archive()` and `add_go_library()` has been updated to use it whenever we're doing a Universal build for MacOS.
- And then we added a new `rustlang.cmake` module to put all of the Rust library building logic into one place, and slowly add in the bits necessary to handle multi-architecture builds. This is structured similarly to the golang module in that we provide a top level `add_rust_library()` function, which internally calls `build_rust_archives()` for each desired architecture.
- `vpnglean` and the `signature` libraries now use `add_rust_library()` to produce their linkable targets.

To produce a universal build, we now need to set the [CMAKE_OSX_ARCHITECTURES](https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_ARCHITECTURES.html) variable. If this value is unset, the default is set to the host architecture of the build machine. A universal build can be produced with:
```
cmake .. -GNinja -DCMAKE_OSX_ARCHITECTURES="arm64;x86_64" -DCMAKE_PREFIX_PATH=<Qt-Path>/lib/cmake
```

We can't really turn on universal builds in CI just yet, since our static builds of Qt only target the x86_64 architecture. They will need to be rebuilt before we can produce official universal builds. But in the meantime, it works just fine with the official Qt libraries from the Qt online installer.

## Reference
Github: #286 ([VPN-82](https://mozilla-hub.atlassian.net/browse/VPN-82))

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [x] I have added thorough tests where needed

### CODE REVIEW

**Comment by brizental** [comment]
Removing myself from the reviewers list as I don't feel qualified to review cmake changes. I looked at them, nothing popped out to me and I learned some new stuff. Thanks for implementing this Owen :) I noticed the Jira ticket has not been added to the December sprint. I think we are expected to do that to all bugs we work on during a sprint.

**Review by mcleinman — CHANGES_REQUESTED** [review]
Getting some errors when building for macOS:
```
error[E0560]: struct `Configuration` has no field named `trim_data_to_registered_pings`
  --> src/lib.rs:42:9
   |
42 |         trim_data_to_registered_pings: false,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Configuration` does not have this field
   |
   = note: available fields are: `upload_enabled`, `data_path`, `application_id`, `max_events`, `delay_ping_lifetime_io` ... and 3 others

error[E0560]: struct `Configuration` has no field named `trim_data_to_registered_pings`
  --> src/lib.rs:79:9
   |
79 |         trim_data_to_registered_pings: false,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Configuration` does not have this field
   |
   = note: available fields are: `upload_enabled`, `data_path`, `application_id`, `max_events`, `delay_ping_lifetime_io` ... and 3 others

For more information about this error, try `rustc --explain E0560`.
error: could not compile `vpnglean` due to 2 previous errors
```
Happy to debug w/ you in real time if you'd like.

**Comment by brizental** [comment]
@mcleinman You need to update the gid submodule. That is the source of your errors :)

File: README.md [hunk 1]
```diff
@@ -237,6 +237,8 @@ Some variables that might be useful when configuring the project:
    identifier. This defaults to `org.mozilla.macos.FirefoxVPN` if not set.
  - `BUILD_VPN_DEVELOPMENT_TEAM=<Development Team ID>`: can be set to change the development
    team used for Xcode certificates. This defaults to `43AQ936H96` if not set.
+ - `CMAKE_OSX_ARCHITECTURES="arm64;x86_64"`: can be set to produce a universal binary that
+   will run on both Intel and Apple silicon devices (experimental).
```
[the remark this answers is not in the document]
Comment (oskirby) [reply README.md:241]: For MacOS, I believe this works on version 6.2.x as well. It's only iOS that requires something newer.

**Review by bakulf — APPROVED** [review]

**Review by mcleinman — APPROVED** [review]
👍🏻 
Cmake is not my specialty, but nothing leaps out at me as off here, and I can successfully get it to build for iOS and macOS. Thanks!

**Pull request merged**